### PR TITLE
Add rebase with merge commit merge style (#3844)

### PR DESCRIPTION
**Pull request by apricote**
This PR adds the `Rebase and Merge (--no-ff)` merge style as requested in #3844.

---

Known Bugs:
- [x] ~~As with the `Rebase and Merge` style, if the rebase can not be done cleanly, the merge will result in a `500` error page~~ Tracked in #4078

### CODE REVIEW

**Comment by apricote** [comment]
I have created a seperate issue to track the error during rebase #4078.

This PR is ready to be reviewed now.

**Comment by apricote** [comment]
@InExtremaRes Currently I am waiting for the review process. You too can leave suggestions/improvements to the code. ;) 

Besides that you can try to convince some maintainers in the discord channel to review it. As the 1.5.0 Release has been cut off and is mostly done, there might be someone willing to do it. 

Edit: @Maintainers How should i proceed with conflicting migrations? As there are always some new ones merged in the one created in the PR is always out of date with the numbering. Should I try to keep up or will you just bump the numbers once you are merging?

**Comment by InExtremaRes** [comment]
@apricote I'd love to review/suggest/improve the code but I don't know _Go_. What is that discord channel you mentioned?

**Comment by techknowlogick** [comment]
@apricote you may find a maintainer that will resolve conflicts for you, however my personal approach is to only touch someones PR when they've asked for help resolving a conflict. My preference to start reviews is that all conflicts are resolved (so that I can trigger drone to build against the most recent commit in master).

Like you said, the best way to get someones attention is to bring it up in the `#develop` channel in discord.

**Comment by techknowlogick** [comment]
@InExtremaRes https://discord.gg/NsatcWJ is the discord server

File: models/migrations/v67.go
Comment (lafriks) [inline models/migrations/v67.go]: I think `AllowRebaseMergeCommit` should be enabled only if `AllowMerge` is already enabled

File: models/pull.go
Comment (lafriks) [inline models/pull.go]: `MergeStyleRebaseMerge` would be better name

**Review by lafriks — CHANGES_REQUESTED** [review]
`Commit` should be removed from all variables/constants

File: models/migrations/v67.go
Comment (apricote) [reply models/migrations/v67.go]: It might even make sense to not enable it at all for existing repositories, as doing so means interfering with the users choosen preferences without giving them the option to skip enabling it.

File: models/pull.go
Comment (apricote) [reply models/pull.go]: Agree, will change it.

File: models/migrations/v67.go
Comment (lunny) [inline models/migrations/v67.go]: ...

**Review by lunny — CHANGES_REQUESTED** [review]

File: models/pull.go
Comment (apricote) [reply models/pull.go]: This works for all places, but the locale file:
```
pulls.rebase_merge_pull_request = Rebase and Merge
pulls.rebase_merge_commit_pull_request = Rebase and Merge (--no-ff)
```
The `MergeStyleRebase` already uses the (potential) name of the new `MergeStyleRebaseMerge`.

Edit:
Also for these locales:
```
settings.pulls.allow_rebase_merge = Enable Rebasing to Merge Commits
settings.pulls.allow_rebase_merge_commit = Enable Rebasing with explicit merge commits (--no-ff)
```

File: models/migrations/v67.go
Comment (apricote) [reply models/migrations/v67.go]: Fixed, I missed the merge artifacts.

**Comment by Claudisimo** [comment]
This seems like a nice feature. It would have been nice to use it a few times that i remember. How is this going?

**Comment by lunny** [comment]
file conflicted.

File: models/migrations/v67.go
Comment (apricote) [reply models/migrations/v67.go]: Changed the migration, `AllowRebaseMerge` is now only set to `true` if all current merge styles are enabled.

File: integrations/pull_merge_test.go
Comment (apricote) [reply integrations/pull_merge_test.go]: Initially I inteded to modify the test case so that an actual rebase is performed and necessary.

The test `TestPullRebase` does not create such a setting, so i just removed the comments.

**Review by lafriks — APPROVED** [review]

**Comment by InExtremaRes** [comment]
@apricote I was testing this branch in our server and I can notice a couple of issues:

1. The message for the merge commit is not customizable.
2. The text `head_repo` is appended to the branch's name at merging.

Looking at the changes it seem like 2. is deliberated ([example](https://github.com/go-gitea/gitea/pull/4052/files#diff-ca95a2152012afcd9192d0ed961e6b9dR409)). Why is that the default behaviour? I think this does not align very well with the default of git, that always preserve the name of the branch.

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
Before @InExtremaRes 's question resolved, don't merge.

**Comment by apricote** [comment]
@InExtremaRes 

1. This behaviour was copied from the `Rebase` style, which did not use/need a commit message. This should be fixed.
2. This behaviour was also copied from the `Rebase` style.
    * `Rebase`: https://github.com/go-gitea/gitea/blob/b387fc147af2078935f51ce35af23e4e3f6e3c7a/models/pull.go#L391
    * `RebaseMerge`: https://github.com/go-gitea/gitea/blob/b387fc147af2078935f51ce35af23e4e3f6e3c7a/models/pull.go#L416
  I do not know why the branch is called like this.

**Comment by lunny** [comment]
maybe @lafriks could know that.

**Comment by InExtremaRes** [comment]
@apricote That's interesting. Well, when you rebase a branch (and `merge --ff` as in default rebase style) there is no commit merge that needs a message and then there's no evidence of that branch's existence anymore, in the commit logs at least. With this PR the message of the commit and by default the name of the branch are both kept in the history so that name could be helpful in some cases.

For the latest I just can guest. I think is for safety: if the rebase/pull/merge/something fails the branch can just be discarded without touching the original one and, as I said, in the Rebase and Squash styles the name of that branch is not seen anymore. But just my guess, and as @lunny said maybe @lafriks  knows better and can explain.

**Comment by InExtremaRes** [comment]
There is another, small issue I didn't realize before. The author of the commit is set to the `Gitea` user instead of the user merging the PR, which is the behavior in the default merge strategy, and presumably the desire one. Is this deliberated for some reason?

**Comment by InExtremaRes** [comment]
@lafriks 

> That is ok that message is not customizable and it is correct that `head_repo` is prepended to branch name for local checkout, that is just not to conflict with exiting branch.

I think I understand the rationales behind, but I would prefer that even if the _name_ of the branch is prepended the commit _message_ would not, since it is clearer in the log what branch was merged. Most of the time the usefulness of this merge style, as opposed to classic rebase style, is that it's straightforward to see what commits are coming from specific branch and are (probably) related or part of the same feature.

@apricote Is it possible to keep the merge commit with the name of the original branch and the author as the user doing the merge? BTW, thanks for this feature, it is working pretty well until now.

**Comment by apricote** [comment]
Build failure seems to be unrelated:

```
package github.com/jteeuwen/go-bindata/...: github.com/jteeuwen/go-bindata/...: invalid import path: malformed import path "github.com/jteeuwen/go-bindata/...": double dot
Makefile:95: recipe for target 'generate' failed
```

---

@InExtremaRes I added the option to set a custom commit title and message. Also the author of the merge commit is now the user doing the merge.

---

The default merge commit message is now the same as with a usual merge:

* **Previous default  message**: `Merge branch 'head_repo_feature-1'`
* **Current default message**: `Merge branch 'feature-1' of apricote/merge-test into master`

**Comment by InExtremaRes** [comment]
Thank you @apricote. I'll be happy to keep testing and report back later. Since I already executed this when the migration version was 74, is there anything I need to do?

**Review by lafriks — APPROVED** [review]

**Comment by apricote** [comment]
@InExtremaRes If you keep using this branch and eventually switch back to the master, you will skip the [Migration v74](https://github.com/go-gitea/gitea/blob/master/models/migrations/v74.go), this migration adds a column (I think) and skipping it will result in a unexpected/invalid database schema.

The migration introduced in this PR is idempotent and can safely be re-run.

You will want to change the current database version to `73` (using standard sql): `UPDATE version SET version = 73 WHERE id = 1;`

On the next server start, this will retrigger all migrations > 73 and your database should be in the desired state.

**Pull request merged**
